File: src/main/java/com/pandus/leetcode/solutions/google/DiameterOfBinaryTree.java
package com.pandus.leetcode.solutions.google;

import com.pandus.leetcode.solutions.model.TreeNode;

public class DiameterOfBinaryTree {
    private int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return ans;
    }

    private int dfs(TreeNode node) {
        if (node == null) return 0;

        int left = dfs(node.getLeft());
        int right = dfs(node.getRight());

        ans = Math.max(ans, left + right);

        return Math.max(left, right) + 1;
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/daily/DiameterOfBinaryTree.kt
package com.pandus.leetcode.solutions.daily

import com.pandus.leetcode.solutions.model.TreeNode
import kotlin.math.max

// Reference: https://leetcode.com/problems/diameter-of-binary-tree
class DiameterOfBinaryTree {

    private var ans = 0
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        dfs(root)
        return ans
    }

    private fun dfs(node: TreeNode?): Int {
        if (node == null) return 0

        val left = dfs(node.left)
        val right = dfs(node.right)

        ans = max(ans, left + right)

        return max(left, right) + 1
    }
}
